pkg/cli/admin/release/info.go

```
@@ -1987,6 +2035,306 @@ func describeBugs(out, errOut io.Writer, diff *ReleaseDiff, dir string, format s
+}
+
+type RpmChangedDiff struct {
+	Old string `json:"old,omitempty"`
+	New string `json:"new,omitempty"`
+}
+
+func (o *InfoOptions) downloadRpmdb(image imagereference.DockerImageReference, dir string) error {
+	// NOTE: in the future, should check if there's an SBOM and get the rpmlist from that instead
+
+	ref := imagesource.TypedImageReference{Type: imagesource.DestinationRegistry, Ref: image}
+
+	tmpDir := dir + ".work"
+	if err := os.RemoveAll(tmpDir); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(tmpDir, 0755); err != nil {
+		return err
+	}
+
+	// We could be a lot more efficient here by filtering the layers which don't
+	// contain the rpmdb. Though we'd still need to make sure we pick up any user-added
+	// layers in case they added more packages.
+
+	opts := extract.NewExtractOptions(genericiooptions.IOStreams{Out: o.Out, ErrOut: o.ErrOut})
+	opts.SecurityOptions = o.SecurityOptions
+	opts.FilterOptions = o.FilterOptions
+	opts.FileDir = o.FileDir
+	opts.ICSPFile = o.ICSPFile
+	opts.IDMSFile = o.IDMSFile
+	opts.PreservePermissions = false
+
+	opts.Mappings = []extract.Mapping{{
+		ImageRef: ref,
+		// ideally we'd have `From: "usr/share/rpm"` here for the machine-os case, but
+		// the extractor can't handle the rpmdb actually being hardlinks to ostree
+		// objects in the tar stream, which would be filtered out
+		To: tmpDir,
+	}}
+
+	if err := opts.Run(); err != nil {
+		return err
+	}
+
+	var rpmdbPath string
+	rpmdbLocations := []string{"var/lib/rpm", "usr/lib/sysimage/rpm", "usr/share/rpm"}
+	for _, location := range rpmdbLocations {
+		tryPath := filepath.Join(tmpDir, location)
+		if fileinfo, err := os.Lstat(tryPath); os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return err
+		} else if fileinfo.IsDir() {
+			rpmdbPath = tryPath
+			break
+		} else {
+			continue
+		}
+	}
+
+	if rpmdbPath == "" {
+		return fmt.Errorf("no rpmdb found in %s", image)
+	}
+
+	if err := os.Rename(rpmdbPath, dir); err != nil {
```

are hardlinks the only indirection?  E.g. if `usr/share/rpm` was a symlink, this `Rename` wouldn't be sufficient, right?  Do we want to use [`os.CopyFS`][1] or something, to at least fail fast if we hit a symlink?

[1]: https://pkg.go.dev/os#CopyFS

---

We use `Lstat` above and only accept directories, so this shouldn't be a concern.